src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,365 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <bsoncxx/document/view.hpp>
+#include <bsoncxx/stdx/optional.hpp>
+#include <mongocxx/stdx.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+class collection;
+
+namespace options {
+
+///
+/// Class representing the optional arguments to a MongoDB create index operation.
+///
+/// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+///
+class MONGOCXX_API index {
+   public:
+    ///
+    /// Base class representing the optional storage engine options for indexes.
+    ///
+    class base_storage_options {
+       public:
+        virtual ~base_storage_options() = default;
+
+       private:
+        friend collection;
+        MONGOCXX_PRIVATE virtual const int type() const = 0;
+    };
+
+    ///
+    /// Class representing the optional WiredTiger storage engine options for indexes.
+    ///
+    class wiredtiger_storage_options : public base_storage_options {
```

Comment:
Note that I have the same ABI concerns about this "sub option" as I do about the overall options class. What is its extensibility story? How can we allow it to describe new options without an ABI break. Obviously, we can extend the class to handle new options when we next do change ABI, but it would be unfortunate if we didn't have a workaround for cases where ABI can't change.